csharp/src/Apache.Arrow/Arrays/PrimitiveArrayBuilder.cs

```diff
@@ -0,0 +1,168 @@
+﻿// Licensed to the Apache Software Foundation (ASF) under one or more
+// contributor license agreements. See the NOTICE file distributed with
+// this work for additional information regarding copyright ownership.
+// The ASF licenses this file to You under the Apache License, Version 2.0
+// (the "License"); you may not use this file except in compliance with
+// the License.  You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+using Apache.Arrow.Memory;
+using System;
+using System.Collections.Generic;
+using System.Linq;
+
+namespace Apache.Arrow
+{
+    public abstract class PrimitiveArrayBuilder<TFrom, TTo, TArray, TBuilder> : IArrowArrayBuilder<TFrom, TArray>
+        where TTo: struct
+        where TArray: IArrowArray
+        where TBuilder: class, IArrowArrayBuilder<TArray>
+    {
+        protected TBuilder Instance => this as TBuilder;
+        protected IArrowArrayBuilder<TTo, TArray, IArrowArrayBuilder<TTo, TArray>> ArrayBuilder { get; }
+
+        internal PrimitiveArrayBuilder(IArrowArrayBuilder<TTo, TArray, IArrowArrayBuilder<TTo, TArray>> builder)
+        {
+            ArrayBuilder = builder ?? throw new ArgumentNullException(nameof(builder));
+        }
+
+        public TArray Build(MemoryPool pool = default) => ArrayBuilder.Build(pool);
+
+        public TBuilder Append(TFrom value)
+        {
+            ArrayBuilder.Append(ConvertTo(value));
+            return Instance;
+        }
+
+        public TBuilder Append(ReadOnlySpan<TFrom> span)
+        {
+            foreach (var value in span)
+            {
+                Append(value);
```

Comment:
>value [](start = 23, length = 5)

It would probably be best to get the `span.Length`, and then ensure capacity is enough for the entire span. That way we only need to allocate and copy at most a single time. Instead of potentially multiple times.